### DISCUSSION TITLE
iTwin viewer slow once deployed

### DISCUSSION THREAD

**Discussion by lijiayi0921** [iTwin/itwinjs-core#3558] [Q&A] [answered]
Hi,

I have a performance issue of my deployement of iTwin viewer.

I implemented iTwin viewer as one page in my application to visualize iModels with custom widgets.
On localhost (dev version) the viewer is fast.
Once deployed, the viewer is much slower. 

After some tests I find out my deployed itwin viewer is way slower on Chrome than on Firefox.
When it happens, the whole page is freezed.

If I summerize (from the fastest to the slowest) : 
Chrome (localhost) > Firefox (localhost) >= Firefox (deployed) >> Chrome (deployed)

Edit : hardware acceleration is enabled on both browsers

And users reported a worse performance after I deployed iTwin 3.1.3.

I also noticed a correlation between the viewer performance and the progress bar speed.
![image](https://user-images.githubusercontent.com/10466395/165522748-362db9fd-f838-4986-bc33-d788228f1117.png)

There is unlikely a server saturation since both memory and cpu on the frontend server are far from being saturated.

I'm aware my description of the problem is very vague and there are too many variables which may be the cause.
But some light would be very welcome !

Thanks in advance

Jiayi

**Comment by pmconne** [comment]
You need to be more specific. For example:
- What things are slow? Tiles loading? Framerate? Property requests? etc etc.
- Details of your deployment.
- If performance was worse after deploying 3.1.3, what version was previously deployed?
- Did you check console for errors?
- Did you monitor network activity for issues?
- etc etc.

**Reply by lijiayi0921** [reply]
Hi @pmconne 

Thanks for taking time to read my post.
I may have found something with your hints.



I checked the network activities and compared the deployed version against localhost version.

The initial phase for opening the viewer takes approximately the same amount of time.

After that, model parts are loaded smoothly in the localhost version (finishes in 3-4 seconds), while in the deployed version, requests to **IModelTileRpcInterface-3.0.0-requestTileTreeProps** are more and more sparse, with the entire page being freezed during the process.

The progress bar never reaches the end, and memory footprint stays lower than the localhost version.

Eventually I could run into an "out of memory" crash for the deployed page. While the localhost page has long finished loading the model and is running smoothly.

Here is a screenshot of the deployed viewer's chrome network activities.

![image](https://user-images.githubusercontent.com/10466395/165565000-134ce8e1-323b-4ca4-b6f2-bd75436e2a46.png)




Below are some more details :


- What things are slow? Tiles loading? Framerate? Property requests? etc etc.
       The whole page (chrome tab) is freezing, including UI elements out of the viewer zone

- Details of your deployment.
      App frontend / backend are deployed as docker images living in cloud servers.
      For localhost version, both front and back are launched in the same windows 10 PC.

- If performance was worse after deploying 3.1.3, what version was previously deployed?
      It was 3.0.1. But with previous versions of iTwin viewer (since 2.14) the performance was already bad when deployed.

- Did you check console for errors?
      Yes. Both deployed and localhost versions have the same 401 errors accessing https://api.bentley.com/realitydata.
      Since the same errors are shared by both environments, I'd suppose they are not the cause.

- Did you monitor network activity for issues?
      Yes. (see above)


Thanks

Jiayi

**Reply by kckst8** [reply]
Hi @lijiayi0921 . Could you please provide the same network activity that you are seeing from localhost?

**Reply by lijiayi0921** [reply]
Hi @kckst8 ,

Here is a screenshot of the network activities for localhost.
Tile requests are not slowing down with the time.

![image](https://user-images.githubusercontent.com/10466395/165740076-eb7074fb-e733-46a2-bd68-0ce44c167f78.png)

The next screenshot is the end of the model loading process. By that time the progress bar reaches the end and dissappears.

![image](https://user-images.githubusercontent.com/10466395/165741122-7348e499-b13d-40e6-9815-b3a1dbeb5457.png)

Thanks in advance !
Jiayi

**Comment by lijiayi0921** [comment]
Hi @pmconne @kckst8 ,
Do you have some idea of how I can adress the issue ?
Many thanks
Jiayi

**Reply by aruniverse** [reply]
Have you looked at flamegraphs? 

Please share with us HAR files for both your local dev version and the deployed version. Please make sure you use the same iModel and you do the exact same actions in the viewer for both sessions.

**Comment by aruniverse** [comment, marked as answer]
After debugging in chrome, we narrowed the issue down to the the package.json `browserlist` was a bit too exhaustive and supporting older browsers and adding too many polyfills and chrome was unable to optimize.

The recommended browserlist is:
```
"browserslist": [
  "last 4 chrome version",
  "last 4 firefox version",
  "last 4 safari version",
  "last 4 ios version",
  "last 4 ChromeAndroid version",
  "last 4 edge version",
  "not dead",
  "not <0.2%"
],
```
